src/main/java/de/fraunhofer/aisec/cpg/passes/CallResolver.java

```diff
@@ -412,21 +543,112 @@ private void applyImplicitCastToArguments(
     }
   }
 
+  /**
+   * Changes the arguments of the ConstructExpression to use the implcit casts instead
+   *
+   * @param constructExpression ConstructExpression
+   * @param implicitCasts Casts
+   */
+  private void applyImplicitCastToArguments(
+      ConstructExpression constructExpression, List<CastExpression> implicitCasts) {
+    if (implicitCasts != null) {
+      for (int i = 0; i < implicitCasts.size(); i++) {
+        if (implicitCasts.get(i) != null) {
+          constructExpression.setArgument(i, implicitCasts.get(i));
+        }
+      }
+    }
+  }
+
+  /**
+   * Resolves a CallExpression to the potential target FunctionDeclarations by checking for ommitted
+   * arguments due to previously defined default arguments
+   *
+   * @param call CallExpression
+   * @return List of FunctionDeclarations that are the target of the CallExpression (will be
+   *     connected with an invokes edge)
+   */
+  private List<FunctionDeclaration> resolveWithDefaultArgs(
+      CallExpression call, List<FunctionDeclaration> initialInvocationCandidates) {
+
+    List<FunctionDeclaration> invocationCandidatesDefaultArgs = new ArrayList<>();
+
+    for (FunctionDeclaration functionDeclaration : initialInvocationCandidates) {
+      addDefaultArgsToCall(functionDeclaration, call);
+      if (functionDeclaration.hasSignature(call.getSignature())) {
+        invocationCandidatesDefaultArgs.add(functionDeclaration);
+      }
+    }
+    return invocationCandidatesDefaultArgs;
+  }
+
+  private List<FunctionDeclaration> resolveWithDefaultArgsFunc(CallExpression call) {
```

Comment:
Why is it different for funcs and methods?

Reply:
In C++ the order of the functions matters, but not the order of the methods. This means a function that is declared below the usage cannot be resolved and the compiler will create an implicit function definition. In those cases we want a dummy function to be the target of the resolved. In methods the definition of a method can be below the usage and it is still correctly resolved.

Reply:
Well... the *declaration* of a method must always be known beforehand. The *definition* can be done later in the code. There is no difference in this regards whether this is a function or a method.

Reply:
There is a difference. Consider the following example:
```
#include <iostream>
using namespace std;
class Base
{
public:
    int calc(int i)
    {
        cout << "int: " << doSmth(i,1) << "\n";
        return i;
    }

    int doSmth(int a, int b) {
        return a+b;
    }
};

int main()
{
    Base base;
    base.calc(1);
    return 0;
}
```
`doSmth` is defined below `calc` but this will compile fine and work as intended

Reply:
Ah, yes. This seems to be a special case that methods of the same class can be used (in inline methods) before they are declared. Pretty sure this only applies to this very special case.

Reply:
http://www.open-std.org/jtc1/sc22/wg21/docs/papers/2013/n3690.pdf

Chapter 3.4.1 specifies this in rather long and boring details. Relevant for this is Paragraph 8. which applies to a "name" used in a function body (within a class). The name must either be declared before its use or it must be a member of that class.

Reply:
The general challenge that I see here that this is getting way to specific for C++ in the end. So I would propose not going into the very peculiar specifics of C++ in this generic class. We might have to tackle this differently in the long run  if we want to be very accurate for specific languages, but this is not our intention at the moment.

What do you think, @konradweiss ?

Reply:
As far as I know we cannot distinguish between methods that are defined within a class or not, so since this case is possible I decided to not enforce the rule that a function/method should be declared before its use when dealing with methods.

I agree with you, that this is very specific and it shouldn't be in this generic class, especially because it makes it very hard to maintain. I already discussed this problem with @konradweiss (see #406). I would suggest to keep these changes for now, as sooner or later we will want to have a more precise call resolution and do the refactoring as soon as possible (until we know how to implement the specific CallResolvers) in a separate PR.

Reply:
Yes, at some point we may have to introduce strategies or LanguageSpecific passes, the question is how to best integrate them such that we can in theory parse multiple classes and therefor have to run/apply multiple strategies/or passes.